src/clj/rems/db/outbox.clj

```diff
@@ -0,0 +1,160 @@
+(ns rems.db.outbox
+  (:require [clj-time.core :as time]
+            [clojure.test :refer [deftest is testing]]
+            [rems.db.core :as db]
+            [rems.db.pg-util :refer [pg-interval->joda-duration joda-duration->pg-interval]]
+            [rems.json :as json]
+            [rems.util :refer [getx]]
+            [schema.coerce :as coerce]
+            [schema.core :as s])
+  (:import [org.joda.time Duration DateTime]))
+
+(def OutboxData
+  {(s/optional-key :outbox/id) s/Int
+   :outbox/type (s/enum :email)
+   :outbox/backoff Duration
+   :outbox/created DateTime
+   :outbox/deadline DateTime
+   :outbox/next-attempt (s/maybe DateTime)
+   :outbox/latest-attempt (s/maybe DateTime)
+   :outbox/latest-error (s/maybe s/Str)
+   s/Keyword s/Any})
+
+(def ^Duration initial-backoff (Duration/standardSeconds 10))
+(def ^Duration max-backoff (Duration/standardHours 12))
+
+(defn put! [data]
+  (let [amended (assoc data
+                       :outbox/created (DateTime/now)
+                       :outbox/next-attempt (DateTime/now)
+                       :outbox/latest-error nil
+                       :outbox/latest-attempt nil
+                       :outbox/backoff initial-backoff)]
+    (s/validate OutboxData amended)
+    (:id (db/put-to-outbox! {:outboxdata (json/generate-string amended)}))))
+
+(defn update! [data]
+  (db/update-outbox! {:id (getx data :outbox/id)
+                      :outboxdata (json/generate-string (dissoc data :outbox/id))}))
+
+
+(def ^:private coerce-outboxdata
+  (coerce/coercer! OutboxData (fn [schema]
+                                (if (= schema Duration)
+                                  #(Duration. (long %))
+                                  (json/coercion-matcher schema)))))
+
+(defn- fix-row-from-db [row]
+  (-> (:outboxdata row)
+      json/parse-string
+      coerce-outboxdata
+      (assoc :outbox/id (:id row))))
+
+(defn- next-attempt-now? [entry]
+  (. (:outbox/next-attempt entry) isBefore (DateTime/now)))
```

Comment:
Next attempt can be nil. It happens when deadline is reached. Possible NPE here?

Also here is an off by one race condition, because the comparison is < instead of <=, so it would not match a newly created entry where next attempt = now.  Adding one millisecond to now or using "not after" would fix it.